File: webapp/Locust/Scripts/app/services/TestSuplementalService.js
(function (angular, undefined) {
    'use strict';

    // Get services module
    var app = angular.module(appName);

    app.service('TestSuplementalService', ['$resource', function ($resource) {

        var noop = angular.noop;
        var resource = $resource('/api/TestSuplemental/:action/:id',
            { id: '@Id' }, {
                Save: {
                    method: 'POST'
                },
                Update: {
                    method: 'PUT'
                },
                Restore: {
                    method: 'POST'
                },
                Delete: {
                    method: 'DELETE',
                    params: {
                        id: "@id"
                    }
                },
                Enable: {
                    method: 'GET',
                    isArray: false
                },
                GetLastTestSuplemental: {
                    method: 'GET',
                    isArray: false
                },
                Get: {
                    method: 'GET',
                    isArray: false
                },
                GetLastOne: {
                    method: 'GET',
                    isArray: false
                },

                GetAll: {
                    method: 'GET',
                    isArray: true
                },

                GetAllSTPByProject: {
                    method: 'GET',
                    isArray: true
                },
                GetForProject: {
                    method: 'GET',
                    isArray: true
                },
                GetForProjectInactives: {
                    method: 'GET',
                    isArray: true
                },
                UpdateNumber: {
                    method: 'PUT'
                },

                GetProject: {
                    method: 'GET',
                    isArray: false
                },
                GetRequirement: {
                    method: 'GET',
                    isArray: false
                },
                GetProjectRequirement: {
                    method: 'GET',
                    isArray: false
                },
                GetProcedures: {
                    method: 'GET',
                    isArray: true
                },
                GetScenarios: {
                    method: 'GET',
                    isArray: true
                },
                AddChangeLog: {
                    method: 'GET',
                    isArray: false
                },
                TestSuplementalChangeLogs: {
                    method: 'GET',
                    isArray: true
                },
                GetForTestProcedure: {
                    method: 'GET',
                    isArray: true
                },
                GetForTestScenario: {
                    method: 'GET',
                    isArray: true
                },
                GetByNumber: {
                    method: 'GET',
                    isArray: false
                }
            }),

            self = this;

        self.GetAll = function (success, error) {
            return resource.GetAll({ action: 'GetAll' }, success || noop, error || noop);
        };
        self.GetAllSTPByProject = function (idProject, success, error) {
            return resource.GetAllSTPByProject({ action: 'GetAllSTPByProject', idProject: idProject }, success || noop, error || noop);
        };
        self.Get = function (idTestSuplemental, success, error) {
            return resource.Get({ action: 'Get', idTestSuplemental: idTestSuplemental }, success || noop, error || noop);
        };
        self.GetForProject = function (idProject, success, error) {
            return resource.GetForProject({ action: 'GetForProject', idProject: idProject }, success || noop, error || noop);
        };
        self.GetForProjectInactives = function (idProject, success, error) {
            return resource.GetForProjectInactives({ action: 'GetForProjectInactives', idProject: idProject }, success || noop, error || noop);
        };
        self.GetLastOne = function (idTestProcedure, idTestScenario, creator, date, success, error) {
            return resource.GetLastOne({ action: 'GetLastOne', idTestProcedure: idTestProcedure, idTestScenario: idTestScenario, creator: creator, date: date }, success || noop, error || noop);
        };
        self.TestSuplementalChangeLogs = function (id, success, error) {
            return resource.TestSuplementalChangeLogs({ action: 'TestSuplementalChangeLogs', id: id }, success || noop, error || noop);
        };
        self.AddChangeLog = function (id, success, error) {
            return resource.AddChangeLog({ action: 'AddChangeLog', id: id }, success || noop, error || noop);
        };
        self.Restore = function (entity, success, error) {
            return resource.Restore({ action: 'Restore' }, entity, success || noop, error || noop);
        };
        self.GetRequirement = function (idtp, idts, success, error) {
            return resource.GetRequirement({ action: 'GetRequirement', idtp: idtp, idts: idts }, success || noop, error || noop);
        };
        self.Delete = function (idTestSuplemental, success, error) {
            return resource.Delete({ action: 'Delete', idTestSuplemental: idTestSuplemental }, success || noop, error || noop);
        };
        self.Enable = function (id, success, error) {
            return resource.Enable({ action: 'Enable', id: id }, success || noop, error || noop);
        };
        self.GetLastTestSuplemental = function (creator, date, success, error) {
            return resource.GetLastTestCase({ action: 'GetLastTestSuplemental', creator: creator, date: date }, success || noop, error || noop)
        };
        self.GetProject = function (idtp, idts, success, error) {
            return resource.GetProject({ action: 'GetProject', idtp: idtp, idts: idts }, success || noop, error || noop)
        };
        self.GetProcedures = function (idstp, success, error) {
            return resource.GetProcedures({ action: 'GetProcedures', idstp: idstp }, success || noop, error || noop)
        };
        self.GetScenarios = function (idstp, success, error) {
            return resource.GetScenarios({ action: 'GetScenarios', idstp: idstp }, success || noop, error || noop)
        };
        self.GetForTestProcedure = function (id, success, error) {
            return resource.GetScenarios({ action: 'GetForTestProcedure', idTp: id }, success || noop, error || noop)
        };
        self.GetForTestScenario = function (id, success, error) {
            return resource.GetScenarios({ action: 'GetForTestScenario', idTs: id }, success || noop, error || noop)
        };
        self.Save = function (entity, success, error) {
            return resource.Save({ action: 'Save' }, entity, success || noop, error || noop);
        };
        self.Update = function (entity, success, error) {
            return resource.Update({ action: 'Update' }, entity, success || noop, error || noop);
        };
        self.UpdateNumber = function (entity, success, error) {
            return resource.UpdateNumber({ action: 'UpdateNumber' }, entity, success || noop, error || noop);
        };

        self.GetByNumber = function (idProject, number, success, error) {
            return resource.GetByNumber({ action: 'GetByNumber', idProject: idProject, number: number }, success || noop, error || noop);
        };
    }]);
})(angular);

File: webapp/Locus.Core/Services/TestSuplementalService.cs
using Locus.Core.DTO;
using Locus.Core.Models;
using Locus.Core.Repositories;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Locus.Core.Services
{
    //Interface which contains methods for each CRUD operation
    public interface ITestSuplementalService
    {
        TestSuplemental Save(TestSuplemental testSuplemental, string user);

        TestSuplemental Update(TestSuplemental testSuplemental, string user);

        TestSuplemental Delete(int idTestSuplemental);

        TestSuplemental Enable(int idTestSuplemental);

        List<TestSuplemental> GetAll();

        TestSuplemental Get(int idTestSuplemental);

        TestSuplemental GetLastOne(int idTestProcedure, int idTestScenario, string creator, string date);

        TestSuplemental GetLastTestSuplemental(string creator, string date);

        TestSuplemental UpdateNumber(TestSuplemental testSuplemental);

        List<SupplementalTestProcedureDTO> GetAllSTPByProject(int idProject);

        Project GetProject(int idtp, int idts);

        Requirement GetRequirement(int idtp, int idts);

        List<TestSuplemental> GetForProject(int idProject);

        List<TestSuplemental> GetForProjectInactives(int idProject);

        List<TestSuplemental> GetForTestProcedure(int tpId);

        List<TestSuplemental> GetForTestScenario(int idTs);

        List<TestProcedure> GetProcedures(int idstp);

        List<TestScenario> GetScenarios(int idstp);

        ChangeLog AddChangeLog(int testsuplementalId, string user);

        List<ChangeLog> TestSuplementalChangeLogs(int id);

        ChangeLog Restore(ChangeLog change_log, string user);

        TestSuplemental GetByNumber(int projectId, string number);
    }
    public class TestSuplementalService : ITestSuplementalService
    {
        private readonly ITestSuplementalRepository _testSuplementalRepository;

        //Constructor of TestCaseService and initialize testCaseRepository
        public TestSuplementalService(ITestSuplementalRepository testSuplementalRepository)
        {
            _testSuplementalRepository = testSuplementalRepository;
        }

        public ChangeLog AddChangeLog(int testsuplementalId, string user)
        {
            return _testSuplementalRepository.AddChangeLog(testsuplementalId, user);
        }

        public TestSuplemental Delete(int idTestSuplemental)
        {
            return _testSuplementalRepository.Delete(idTestSuplemental);
        }

        public TestSuplemental Enable(int idTestSuplemental)
        {
            return _testSuplementalRepository.Enable(idTestSuplemental);
        }

        public TestSuplemental Get(int idTestSuplemental)
        {
            return _testSuplementalRepository.Get(idTestSuplemental);
        }

        public List<TestSuplemental> GetAll()
        {
            return _testSuplementalRepository.GetAll();
        }

        public List<SupplementalTestProcedureDTO> GetAllSTPByProject(int idProject)
        {
            return _testSuplementalRepository.GetAllSTPByProject(idProject);
        }

        public TestSuplemental GetByNumber(int projectId, string number)
        {
            return _testSuplementalRepository.GetByNumber(projectId, number);
        }

        public List<TestSuplemental> GetForProject(int idProject)
        {
            return _testSuplementalRepository.GetForProject(idProject);
        }

        public List<TestSuplemental> GetForProjectInactives(int idProject)
        {
            return _testSuplementalRepository.GetForProjectInactives(idProject);
        }

        public List<TestSuplemental> GetForTestProcedure(int tpId)
        {
            return _testSuplementalRepository.GetForTestProcedure(tpId);
        }

        public List<TestSuplemental> GetForTestScenario(int idTs)
        {
            return _testSuplementalRepository.GetForTestScenario(idTs);
        }

        public TestSuplemental GetLastOne(int idTestProcedure, int idTestScenario, string creator, string date)
        {
            return _testSuplementalRepository.GetLastOne(idTestProcedure, idTestScenario, creator, date);
        }

        public TestSuplemental GetLastTestSuplemental(string creator, string date)
        {
            return _testSuplementalRepository.GetLastTestSuplemental(creator, date);
        }

        public List<TestProcedure> GetProcedures(int idstp)
        {
            return _testSuplementalRepository.GetProcedures(idstp);
        }

        public Project GetProject(int idtp, int idts)
        {
            return _testSuplementalRepository.GetProject(idtp, idts);
        }

        public Requirement GetRequirement(int idtp, int idts)
        {
            return _testSuplementalRepository.GetRequirement(idtp, idts);
        }

        public List<TestScenario> GetScenarios(int idstp)
        {
            return _testSuplementalRepository.GetScenarios(idstp);
        }

        public ChangeLog Restore(ChangeLog change_log, string user)
        {
            return _testSuplementalRepository.Restore(change_log, user);
        }

        public TestSuplemental Save(TestSuplemental testSuplemental, string user)
        {
            return _testSuplementalRepository.Save(testSuplemental, user);
        }

        public List<ChangeLog> TestSuplementalChangeLogs(int id)
        {
            return _testSuplementalRepository.TestSuplementalChangeLogs(id);
        }

        public TestSuplemental Update(TestSuplemental testSuplemental, string user)
        {
            return _testSuplementalRepository.Update(testSuplemental, user);
        }

        public TestSuplemental UpdateNumber(TestSuplemental testSuplemental)
        {
            return _testSuplementalRepository.UpdateNumber(testSuplemental);
        }
    }
}
